api/src/main/java/org/pmiops/workbench/api/DataSetController.java

```diff
@@ -438,6 +438,8 @@ private void formatTimestampValues(List<DataSetPreviewValueList> valuePreviewLis
   @Override
   public ResponseEntity<EmptyResponse> exportToNotebook(
       String workspaceNamespace, String workspaceId, DataSetExportRequest dataSetExportRequest) {
+    workspaceService.requireActiveBilling(workspaceNamespace, workspaceId);
```

Comment:
These checks should follow the access check, else this reveals some amount of information about existence of the workspace and its billing status, to any caller. In the cases where there is no explicit access check happening in the controller method, it would be probably be *correct* to add one - or else push this logic down into the service which is probably making that check. If this gets too hairy, I could probably be convinced it's not worth the effort for some of these without heavier refactoring.

Reply:
The need to add code to a number methods in just the right place makes me think we've missed an abstraction/opportunity to have something more systematic in place. Maybe a shim in between the ApiDelegate class and the Controller would be a good place to stick these validation methods generically.

What's here is probably enough/appropriate for now to avoid scope creep, but if we need multiple overlapping kinds of checks soon, it could get unwieldy.

Reply:
I was able to correctly order the access checks by pushing down the validate call to the underlying copy notebook function which performs the workspace access check. One side effect of this change, is that Rename Notebook will no longer be allowed since rename works by creating a new file and deleting the original.

Given that we were unsure if Rename should be allowed or not, I thought it would be OK to not allow it if it makes the implementation easier.

Another note to Jay's point, this refactoring went in a direction of checking access at the point where we perform the relevant action instead of checking it at the top of the controller methods. It's a design that he brought up a while ago and I think that may be the way to go if we need to add more access control. It is going to be difficult to have to fully understand what an API call is going to do and declare the necessary permissions.

One downside to that approach though, is that our access control checks can be expensive (if we ask Firecloud) so we do not want to be calling it multiple times in a single client serving API call. One solution to this may be to add some caching mechanism that exists for the duration of a client serving API call so that we can add access control checks at various points without worrying about performance.